tests/unit/classifier/svm/LibLinear_unittest.cc

```diff
@@ -6,1320 +6,480 @@
 
-	ll->set_liblinear_solver_type(liblinear_solver_type);
-	ll->train();
-	pred = ll->apply_binary(test_feats);
-	SG_REF(pred);
+		if (l1)
+				generate_data_l1_simple();
+		else
+				generate_data_l2_simple();
 
-	liblin_accuracy = eval->evaluate(pred, ground_truth);
+		auto ll = new CLibLinear();
+		SG_REF(ll);
 
-	EXPECT_NEAR(liblin_accuracy, 1.0, 1e-6);
 
-	SG_UNREF(ll);
-	SG_UNREF(train_feats);
-	SG_UNREF(test_feats);
-	SG_UNREF(ground_truth);
-	SG_UNREF(eval);
-	SG_UNREF(pred);
-}
+		float64_t liblin_accuracy;
+		auto eval = new CContingencyTableEvaluation();
+		SG_REF(eval);
 
-TEST(LibLinear,train_L2R_L2LOSS_SVC_DUAL)
-{
-	LIBLINEAR_SOLVER_TYPE liblinear_solver_type = L2R_L2LOSS_SVC_DUAL;
 
-	CDenseFeatures<float64_t>* train_feats = NULL;
-	CDenseFeatures<float64_t>* test_feats = NULL;
-	CBinaryLabels* ground_truth = NULL;
 
-	generate_data_l2(train_feats, test_feats, ground_truth);
+		ll->set_bias_enabled(biasEnable);
+		ll->set_features(train_feats);
+		if (C_value)
+				ll->set_C(0.1,0.1); //Only in the case of L2R_L1LOSS_SVC_DUAL
+		ll->set_labels(ground_truth);
+		ll->set_liblinear_solver_type(liblinear_solver_type);
+		ll->train();
 
-	CLibLinear* ll = new CLibLinear();
+		auto pred = ll->apply_binary(test_feats);
+		SG_REF(pred);
+		liblin_accuracy = eval->evaluate(pred, ground_truth);
 
-	CBinaryLabels* pred = NULL;
-	float64_t liblin_accuracy;
-	CContingencyTableEvaluation* eval = new CContingencyTableEvaluation();
+		if (!biasEnable)
+		{
+			for(int i=0;i<t_w.vlen;i++)
+				EXPECT_NEAR(ll->get_w()[i], t_w[i], 1e-5);
+		}
+		else
+		{
+			for(int i=0;i<t_w.vlen;i++)
+				EXPECT_NEAR(ll->get_w()[i], t_w[i], 4e-5);
+			EXPECT_NEAR(ll->get_bias(), t_w[2], 4e-5);
+		}
+		EXPECT_NEAR(liblin_accuracy, 1.0, 1e-5);
 
-	ll->set_bias_enabled(false);
-	ll->set_features(train_feats);
-	ll->set_labels(ground_truth);
+		SG_UNREF(ll);
+		SG_UNREF(eval);
+		SG_UNREF(pred);
+	}
 
-	ll->set_liblinear_solver_type(liblinear_solver_type);
-	ll->train();
-	pred = ll->apply_binary(test_feats);
-	SG_REF(pred);
-	liblin_accuracy = eval->evaluate(pred, ground_truth);
+protected:
+	void generate_data_l1()
+	{
+		generate_data("L1");
+	}
+	void generate_data_l2()
+	{
+		generate_data("L2");
+	}
+	void generate_data_l1_simple()
+	{
+		generate_data_simple("L1_SIMPLE");
+	}
+	void generate_data_l2_simple()
+	{
+			generate_data_simple("L2_SIMPLE");
+	}
+	void generate_data(std::string type) //Type either "L1" or "L2"
+	{
+		/*
+			First part is shared by l1 and l2
+		*/
 
-	EXPECT_NEAR(liblin_accuracy, 1.0, 1e-6);
+		index_t num_samples = 50;
+		sg_rand->set_seed(5);
 
-	SG_UNREF(ll);
-	SG_UNREF(train_feats);
-	SG_UNREF(test_feats);
-	SG_UNREF(ground_truth);
-	SG_UNREF(eval);
-	SG_UNREF(pred);
-}
+		SGMatrix<float64_t> data =
+			CDataGenerator::generate_gaussians(num_samples, 2, 2);
+		CDenseFeatures<float64_t> features(data);
 
-TEST(LibLinear,train_L2R_L2LOSS_SVC)
-{
-	LIBLINEAR_SOLVER_TYPE liblinear_solver_type = L2R_L2LOSS_SVC;
+		/*
```

Comment:
the formatting of this comment is weird. Also the comment is superflous: if you name your variables nicely (they are) then the code is self explaining and we dont need this comment as it doesnt add any information